### DISCUSSION TITLE
STM32G0B1RE dual can configuration

### DISCUSSION THREAD

**Discussion by adri1mart1** [zephyrproject-rtos/zephyr#67444] [Q&A] [answered]
Hello,

I am currently creating a dts file for a custom board based on an STM32G0B1RE chip that has a dual FDCAN controller (See: https://www.st.com/en/microcontrollers-microprocessors/stm32g0b1re.html ).

However, digging into different STM32G0B1RE dts related file to find nodes, I could only find a single fdcan1 node here --> https://github.com/zephyrproject-rtos/zephyr/blob/23b3e5741ea39543f29efaadaa53c21c072ad514/dts/arm/st/g0/stm32g0b1.dtsi#L33C5-L33C5 

Is there any reason why fdcan2 is not there ? 

(Searching everywhere...)
```
zephyr/dts/arm/st/g0$ grep -nre "fdcan"
stm32g0b1.dtsi:33:              fdcan1: can@40006400 {
stm32g0b1.dtsi:34:                      compatible = "st,stm32-fdcan";
```

**Comment by ndrs-pst** [comment]
@adri1mart1, After reviewing the reference manual, it appears that `FDCAN1` and `FDCAN2` indeed share the same `IRQ`. The distinction between the interrupt sources requires `SYSCFG_ITLINE21` and `SYSCFG_ITLINE22`. Therefore, my current understanding is that effectively supporting multiple CANFDs that share the same IRQ presents a significant challenge, one that hasn't been implemented yet since it's involved with IRQ handling in architecture level (isr_wrapper.S).

By the way, if you want to test `fdcan2` while disabled `fdcan1` then you may add this to .overlay file.

```
fdcan2: can@40006800 {
    compatible = "st,stm32-fdcan";
    reg = <0x40006800 0x400>, <0x4000b800 0x350>;
    reg-names = "m_can", "message_ram";
    interrupts = <21 0>, <22 0>;
    interrupt-names = "LINE_0", "LINE_1";
    clocks = <&rcc STM32_CLOCK_BUS_APB1 0x00001000>;
    bosch,mram-cfg = <0x0 28 8 3 3 0 3 3>;
    sample-point = <875>;
    sample-point-data = <875>;
    status = "okay";
};
```

**Reply by henrikbrixandersen** [reply]
Actually, Zephyr supports shared interrupts on the architecture level now, so it should be possible to implement support for both FDCAN instances.

**Reply by ndrs-pst** [reply]
@henrikbrixandersen, Thanks for the informations.

@adri1mart1 You may have to take a look at these issues 
https://github.com/zephyrproject-rtos/zephyr/issues/67087
https://github.com/zephyrproject-rtos/zephyr/issues/49379

**Reply by adri1mart1** [reply]
Thank you for your fast answers. 

As far as I've read issues / PRs, the only remaining step would be to modify the can_stm32_fdcan.c driver to support shared irq that has been recently merged.

There is this PR https://github.com/zephyrproject-rtos/zephyr/pull/34963 that seems very linked with what I need, however, it was opened in May 2021 and left as is.

I'm not sure I'll be able to modify the driver myself, do you suggest me to opened a new issue to highlight exactly what I would ideally like ?

**Reply by ndrs-pst** [reply, marked as answer]
@adri1mart1, I've taken a look at PR [#34963](https://github.com/zephyrproject-rtos/zephyr/pull/34963), and it aligns closely with the requirements for supporting shared IRQ in the `can_stm32_fdcan.c` driver. Since it has been inactive for a while, I encourage you to reference this PR and try to implement the necessary features 🚀.

There seems to be no need to open a new issue. Instead, you could return with a new PR that builds on the original one, giving credit to the initial work. ✌️

**Reply by adri1mart1** [reply]
Alright, thanks, I'll have a look and give any feedback if needed. I'll mark your comment as an answer.
Regards,
Adrien M.
